### PR TITLE
feat(recompute): add mamba to recompute_modules

### PR DESCRIPTION
**Pull request by yuchenwang3**
- [x] I, the PR author, have personally reviewed every line of this PR.

## What

Add `mamba` as a selective activation-recompute target. With `--recompute-granularity selective --recompute-modules '[mamba]'`, `MambaLayer` checkpoints the mixer (convolution plus selective SSM/SSD compute) instead of retaining its activations.

## Why

In Mamba-hybrid models the mixer activations can be a material part of long-sequence memory. Megatron already lets selective recompute target attention, MLP, MoE, and several GDN components, but not Mamba, leaving full-layer recompute as the only way to trade compute for this activation memory.

## Implementation

- Add `mamba` to the `recompute_modules` validation and documentation.
- Use `tensor_parallel.checkpoint` for BF16/FP16/FP32.
- Use Transformer Engine checkpointing for FP8/FP4 so recomputation does not update amax/scaling history twice.
- Preserve `inference_context` and `packed_seq_params` through both paths.
- Keep eval and inference on the direct mixer path.
- Bypass activation checkpointing during local CUDA-graph warmup/capture, matching the existing tensor-parallel checkpoint contract and avoiding TE checkpoint's `.grad()` incompatibility.

The option is default-off and does not change existing runs unless `mamba` is explicitly selected.

## Tests

The focused suite covers disabled/not-selected/eval/inference bypass, BF16 routing, FP8 and FP4 TE routing, packed-sequence argument propagation, and FP8/FP4 CUDA-graph warmup/capture bypass. It passes 11 tests on the rebased branch. Black, isort, Ruff, Pylint, `py_compile`, and `git diff --check` also pass.

An 8×B200 run on the exact PR head covered packed BF16 and FP8 training plus local Mamba CUDA-graph capture/replay. At sequence length 16K (CP1/DP8), selective and full FP8 recompute both completed without NaNs or OOMs and differed by about `1e-5` relative loss. A 20-step BF16 selective/full comparison had a maximum absolute difference of `2.37e-3`; repeating selective with the same seed produced `2.24e-3` run-to-run variation, so the cross-mode difference was within the measured baseline nondeterminism.

At the 24K BF16 boundary with the same model/topology, no recompute OOMed before completing its first step, while selective Mamba recompute completed both steps without NaNs (`10.58632 → 10.09310`) with rank-0 peak allocated memory of 96.13 GiB and 102.09 GiB. At 32K both modes OOMed before a complete step. Because the failed baseline has no common completed step, this is capacity evidence rather than a like-for-like percentage-memory comparison.

The intended CP8 matrix could not run because this environment did not expose a compatible deterministic attention backend, so these results are explicitly CP1/DP8 rather than CP8. Raw numbers and the CUDA-graph baseline comparison are in the PR discussion.

### CODE REVIEW

File: megatron/core/ssm/mamba_layer.py [hunk 1]
```diff
@@ -152,9 +164,25 @@ def forward(
...
         hidden_states = apply_module(self.norm)(hidden_states)
 
-        mixer_out_with_bias = self.mixer(
-            hidden_states, inference_context=inference_context, packed_seq_params=packed_seq_params
-        )
+        if self.recompute_mamba_mixer and self.training and inference_context is None:
+            # Recompute the mixer during the backward pass to save activation memory.
+            # Guarded to training only (inference_context is None) so the stateful SSM
+            # update is never re-executed during inference.
+            mixer_out_with_bias = tensor_parallel.checkpoint(
```
[the remark this answers is not in the document]
Comment (yuchenwang3) [reply megatron/core/ssm/mamba_layer.py]: Addressed in 20eeb77 — mirrored the fp8/fp4 `te_checkpoint` branch from `transformer_layer.py` (`MambaLayer` now keeps `pg_collection` for the TP group). bf16 path unchanged. The planned convergence/peak-memory validation will include an fp8 run to exercise this path.

**Comment by yuchenwang3** [comment]
@Phlip79 Thanks for the quick vetting and the strict review. The fp8/fp4 finding is addressed in 20eeb77 (mirrors the `te_checkpoint` branch from `transformer_layer.py`). Could you `/ok to test 20eeb77` so CI runs on the updated SHA?

**Comment by yuchenwang3** [comment]
@Phlip79 following up on my Jul 10 comment — the fp8/fp4 gating your strict-review flagged is fixed in 20eeb77, and that SHA is still waiting on `/ok to test` (only DCO and check-approval have run). Could you authorize CI on it so this can move to review?

**Comment by yuchenwang3** [comment]
@Phlip79 rebased this onto current main as 4811efa0a01cb7f0b6f5d1e3081a902298199be5. The FP8/FP4 TE path from your earlier review is preserved, and the refreshed branch also handles local CUDA-graph warmup/capture explicitly; without that bypass, TE checkpoint can fail on the graph builder’s `.grad()` path. The focused suite now has 11 passing routing/bypass cases and the 8×B200 matrix is in progress. Could you review this SHA and run `/ok to test 4811efa0a01cb7f0b6f5d1e3081a902298199be5`?

**Comment by yuchenwang3** [comment]
Finished the 8×B200 pass on `4811efa`.

The focused recompute suite is 11/11. For packed CP1/DP8 training at seq 16K, BF16 selective and full both completed 20 steps with no NaNs or OOMs. Their max absolute loss difference was `2.37e-3`; a same-seed selective repeat differed from the first selective run by `2.24e-3`, so the selective/full delta is within the run-to-run variation measured on this setup.

FP8 selective and full also completed cleanly. The two losses were `10.58814, 10.52099` for selective and `10.58803, 10.52089` for full (roughly `1e-5` relative difference). Rank-0 peak allocated memory was 66.26 GB for selective and 30.90 GB for full. In BF16, selective reached 67.30 GB and full 30.65 GB by the second step; the no-recompute two-step baseline was about 71.4 GB.

I also ran FP8 selective recompute with local Mamba CUDA graphs. The non-packed case built 40 graphs and completed 4/4 steps; the post-capture steps ran in about 136 ms and 128 ms, so replay was exercised rather than only capture. The packed case built the same 40 graphs but failed the first runner lookup. The identical failure reproduces with recompute disabled, so that is a packed/dynamic-runner baseline limitation in this configuration, not selective recompute behavior.

The intended CP8 run was blocked because this container had no compatible deterministic attention backend for CP8. I am reporting the topology explicitly rather than presenting the CP1/DP8 fallback as CP8 evidence.

**Comment by yuchenwang3** [comment]
One useful boundary point to add: with the same packed BF16 CP1/DP8 setup at sequence length 24K, the no-recompute run OOMed before completing its first step, while selective Mamba recompute completed both steps with no NaNs (`10.58632 → 10.09310`). Rank-0 peak allocated memory was 96.13 GiB on step 1 and 102.09 GiB on step 2.

At 32K both modes OOMed before a complete step. Since the 24K baseline has no completed step, I am treating this as evidence that selective recompute extends the trainable sequence-length boundary, not as a like-for-like percentage-memory comparison.

**Comment by yuchenwang3** [comment]
Re-ran the focused recompute suite on current head `caee7ca7fecf68b5b086953209a3f902dc3ed5f2` with 8 ranks:

```
python -m pytest -q -p no:cacheprovider tests/unit_tests/ssm/test_mamba_recompute.py
```

Each rank executed all 12 cases: 12 passed, 0 skipped, 0 failed (96 passes total). This covers packed-sequence metadata forwarding and the FP8/FP4 recompute routing, and the worktree was clean.

This is unit-level validation only; I am still keeping the separate production Mamba forward/backward off-vs-selective gradient parity run as a distinct follow-up rather than treating these unit results as that evidence. Could you `/ok to test caee7ca7fecf68b5b086953209a3f902dc3ed5f2` so CI can run on the current head?

**Comment by yuchenwang3** [comment]
@NVIDIA/mcore-oncall — production forward/backward parity is now complete on head `caee7ca7fecf68b5b086953209a3f902dc3ed5f2` across 8×B200.

The run exercised the actual `MambaLayer -> MambaMixer -> ssd_combined` CUDA path with recompute disabled versus `recompute_modules=["mamba"]`, restoring identical RNG immediately before each forward. All 8 ranks passed with 0 skips/failures; outputs, input gradients, and all 12 parameter gradients had max absolute and relative difference `0`, with no NaN/Inf/OOM. `causal-conv1d==1.7.0` was importable and its fused forward callable was active. The source tree was clean and the distributed command exited 0.

This closes the production-parity follow-up from my previous comment. Could you route this for review and `/ok to test caee7ca7fecf68b5b086953209a3f902dc3ed5f2`?

**Comment by yuchenwang3** [comment]
@Phlip79, could you revisit this selective Mamba recompute PR? The earlier FP8/FP4 routing concern has been addressed, and the follow-up comments document both targeted tests and the separate production forward/backward parity check. Could you review the current diff or route it to the appropriate owner, and authorize CI if appropriate? Thanks.

**Comment by yuchenwang3** [comment]
@jiemingz Rebased at `ebe20fd8041`, preserving the new two-stage Mamba methods and the selective-recompute helper. Local lint checks pass; GPU parity and the two-stage/recompute combination still need validation on this head. Could you review the integration and authorize CI?
